File: packages/shared/mongodb/Purchase.ts
import type {
  Category,
  ClaimType,
  IngestionSource,
  ISODateString,
  Platform,
  PurchaseDateBasis,
  PurchaseStatus,
  UUID,
} from "./types";

export interface ExtractionConfidence {
  platform: number;
  price: number;
  overall_min: number;
  order_id?: number | null;
  product_name?: number | null;
  product_id?: number | null;
  price_paid?: number | null;
  member_price_at_purchase?: number | null;
  purchase_date?: number | null;
  member_tier_at_purchase?: number | null;
  variant?: number | null;
  category?: number | null;
}

export interface Purchase {
  _id: UUID;
  updated_at: ISODateString | null;
  user_id: UUID;
  platform: Platform;
  category: Category;
  product_name: string;
  product_id: string;
  product_url: string | null;
  variant: string | null;
  // Category-specific fields. Null for retail; populated for airline/hotel.
  fare_class: string | null;
  room_type: string | null;
  bed_type: string | null;
  rate_type: string | null;
  price_paid: number;
  member_price_at_purchase: number | null;
  non_member_price_at_purchase: number | null;
  currency: "USD";
  purchase_date: ISODateString;
  purchase_date_basis: PurchaseDateBasis;
  window_expires: ISODateString;
  order_id: string;
  member_tier_at_purchase: string | null;
  status: PurchaseStatus;
  claim_type: ClaimType;
  monitoring_cadence_minutes: number;
  last_checked_at: ISODateString | null;
  ingested_at: ISODateString;
  ingestion_source: IngestionSource;
  receipt_storage_url: string | null;
  receipt_hash: string | null;
  extraction_confidence: ExtractionConfidence;
}

File: packages/shared/mongodb/purchase.py
"""Purchase collection — mirror of Purchase.ts."""

from pydantic import BaseModel

from .base import BaseDocument
from .enums import (
    Category,
    ClaimType,
    IngestionSource,
    PurchaseDateBasis,
    PurchaseStatus,
)


class ExtractionConfidence(BaseModel):
    platform: float
    price: float
    overall_min: float


class Purchase(BaseDocument):
    user_id: str
    platform: str
    category: Category
    product_name: str
    product_id: str
    product_url: str | None
    variant: str | None
    fare_class: str | None
    room_type: str | None
    bed_type: str | None
    rate_type: str | None
    price_paid: float
    member_price_at_purchase: float | None
    non_member_price_at_purchase: float | None
    currency: str
    purchase_date: str
    purchase_date_basis: PurchaseDateBasis
    window_expires: str
    order_id: str
    member_tier_at_purchase: str | None
    status: PurchaseStatus
    claim_type: ClaimType
    monitoring_cadence_minutes: int
    ingested_at: str
    ingestion_source: IngestionSource
    receipt_storage_url: str | None
    receipt_hash: str | None
    extraction_confidence: ExtractionConfidence
